### ISSUE TITLE
[no-unnecessary-condition] False positive unnecessary optional chain on nested `void` values

### ISSUE THREAD

**Issue by lukekarrys** [typescript-eslint/typescript-eslint#9754]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

Given the following code, I would expect it to lint successfully:

```typescript
const maybe = <T>(v: T): T | void => v

const a = maybe({
  key1: {
    key2: maybe({
      i: 1,
    }),
  },
})?.key1.key2?.i
//           ^
// 9:14  error  Unnecessary optional chain on a non-nullish value  @typescript-eslint/no-unnecessary-condition
```

In my reproduction the second optional chain is removed if I run with `--fix` but then TypeScript errors with the following:

```
> tsc --noEmit

index.ts:9:15 - error TS2339: Property 'i' does not exist on type 'void | { i: number; }'.
  Property 'i' does not exist on type 'void'.

9 })?.key1.key2.i
                ~


Found 1 error in index.ts:9
```

I also want to note that this behavior does not occur if the value could be `null` only `void`. It successfully lints if I change the return of `maybe` to `T | null`.

In my reproduction I also tried to narrow it down as far as I could and found that these examples lint successfully:

```ts
// This lints fine
const b = maybe({
  key1: maybe({
    i: 1,
  }),
})?.key1?.i

// So does this
const c = {
  key1: {
    key2: maybe({
      i: 1,
    }),
  },
}.key1.key2?.i
```

### Reproduction Repository Link

https://gist.github.com/lukekarrys/979ea0dd68b52937e9a257b4d2c9c293

### Repro Steps

```sh
git clone https://gist.github.com/lukekarrys/979ea0dd68b52937e9a257b4d2c9c293
npm install
npm run lint
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.0.1` |
| `TypeScript`                           | `5.5.4` |
| `ESLint`                               | `9.8.0` |
| `node`                                 | `22.4.0` |

**Comment by Josh-Cena** [comment]
TIL that TypeScript allows optional chaining access on `void` 😅

Not sure this code is good to be allowed though, since `void` could be anything, including a non-null value. This code is just unsound.

**Comment by bradzacher** [comment]
It doesn't look like this reproduces in [our playground](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=MYewdgzgLgBAtgQwJ4CMCmMC8MA8AVAPgAoA3ALhjwEoK8YAfGEkASwBMsCmAob0SWAizxk6IgG9uMGAGs0SAIwVJ06XKQAmColRoJU1dJYUFAGgPSAvlXNXz1gPwA6dQpfyNzltwD0PwwHSAH5Bvv4AnGQKACzSaABO8SDx0gCqYGBowGgQEAjxSDAgAA5QLOAIADYwwAAWCCxgRU1CYOAAtGAArpWVLBC1TFVdGDAAAlBIxTnA8Syl7Tl9YFA%2Bbe1dGVk5eQXt-GwsZeC8fpS1-TDLUBAwAGaNaHzg0DAowjpiKrLySiK6%2BkMxhgZgM1nsVGcri8p38AGUQDA2CAcjAoBcIM8BDVhN9XMoLD9NNpRHpvkCTLZVOCwfZ3Ip6Z4nN4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by lukekarrys** [comment]
> It doesn't look like this reproduces in [our playground](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=MYewdgzgLgBAtgQwJ4CMCmMC8MA8AVAPgAoA3ALhjwEoK8YAfGEkASwBMsCmAob0SWAizxk6IgG9uMGAGs0SAIwVJ06XKQAmColRoJU1dJYUFAGgPSAvlXNXz1gPwA6dQpfyNzltwD0PwwHSAH5Bvv4AnGQKACzSaABO8SDx0gCqYGBowGgQEAjxSDAgAA5QLOAIADYwwAAWCCxgRU1CYOAAtGAArpWVLBC1TFVdGDAAAlBIxTnA8Syl7Tl9YFA%2Bbe1dGVk5eQXt-GwsZeC8fpS1-TDLUBAwAGaNaHzg0DAowjpiKrLySiK6%2BkMxhgZgM1nsVGcri8p38AGUQDA2CAcjAoBcIM8BDVhN9XMoLD9NNpRHpvkCTLZVOCwfZ3Ip6Z4nN4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

I originally wasn't able to reproduce it in the playground either, which is why I created the gist. But weirdly enough the link you just posted does reproduce for me now.

<img width="1137" alt="Screenshot 2024-08-07 at 11 43 49 PM" src="https://github.com/user-attachments/assets/f9295567-ff14-49be-ac14-1df76888ee19">

**Comment by bradzacher** [comment]
Something weird.... it wasn't working and then I open my link and it repros 😅
something must have been borked somewhere.. But it does repro at least!

**Comment by lukekarrys** [comment]
> TIL that TypeScript allows optional chaining access on `void` 😅
> 
> Not sure this code is good to be allowed though, since `void` could be anything, including a non-null value. This code is just unsound.

I agree that the example code is unsound and my example code is quite convoluted. Here is the slightly more real world example that led me to create this reproduction:

```typescript
const maybeGenerator = () => {
  if (Math.random()) {
    return
  }
  return {
    *g() {
      yield { i: 1 }
    },
  }
}

maybeGenerator()?.g().next().value?.i
//                                ~~
//  12:35  error  Unnecessary optional chain on a non-nullish value  @typescript-eslint/no-unnecessary-condition
```

**Comment by bradzacher** [comment]
> Not sure this code is good to be allowed though, since void could be anything, including a non-null value. This code is just unsound.

I'd definitely agree with Josh here this code is very unsound!
`void` is akin to `any` when used in a return type location.

Really we should treat it like `any` in the rule (i.e. we just ignore optional chains off `any`).

**Comment by lukekarrys** [comment]
> Really we should treat it like any in the rule (i.e. we just ignore optional chains off any).

That sounds good to me. My only annoyance with the current behavior is that it will fix by removing the optional chain, but then TypeScript will error instead.

So treating it the same as `any` will fix that part.

**Comment by lukekarrys** [comment]
I did find this test case which asserts `void` is allowed currently: https://github.com/typescript-eslint/typescript-eslint/blob/a26e3c7056fd1f64ebce6eb9a9d9006e10e2bc8d/packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts#L416-L417

After finding that I experimented a little more and found that it only happens when the root object is nullish and the nested property could be `void`: [playground link](https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=C4TwDgpgBAalC8UDOwBOBLAdgcygHygDcB7dAEwCgAbCYIhKAcgEMAjAY0amaVgG4KhAPwA6ANoAGALoCadQgCYGAb3oBfbrwCumMhABmWCGXxRVhAFywoawQtHDx0ikA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

For consistency (if changing the current test cases would be considered a regression) could nested properties be treated the same way as the test case?

**Comment by lukekarrys** [comment]
@bradzacher Based on my last comment, do you have an opinion on if the fix should be:
1. Make `void` act the same as `any`, and  alter the existing test case
2. Make `void` act the same as it does on a value on a nested property of a `null|undefined` object?

I’d like to attempt a PR for this.

**Comment by Josh-Cena** [comment]
+1 to make it behave as `any`

**Comment by bradzacher** [comment]
Ideally - both fixes! 😁


Void should generally be treated as any because it's unsafe to assume anything but that.

But also as you pointed out it looks like void is not correctly handled in this case when the chaining is off a nullish object.
